kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -0,0 +1,331 @@
+   * written to {@code asyncResponse} asynchronously.
+   */
+  public final void resume(AsyncResponse asyncResponse) {
+    AsyncResponseQueue responseQueue = new AsyncResponseQueue();
+    responseQueue.asyncResume(asyncResponse);
+    while (hasNext()) {
+      responseQueue.push(next().handle(this::handleNext));
+    }
+    responseQueue.close();
+  }
+
+  private ResultOrError handleNext(T result, @Nullable Throwable error) {
+    if (error == null) {
+      return ResultOrError.result(result);
+    } else {
+      log.debug("Error processing streaming operation.", error);
+      return ResultOrError.error(EXCEPTION_MAPPER.toErrorResponse(error.getCause()));
+    }
+  }
+
+  abstract boolean hasNext();
+
+  abstract CompletableFuture<T> next();
+
+  private static final class InputStreamingResponse<T> extends StreamingResponse<T> {
+    private final MappingIterator<T> input;
+
+    private InputStreamingResponse(MappingIterator<T> input) {
+      this.input = requireNonNull(input);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return input.hasNext();
+    }
+
+    @Override
+    public CompletableFuture<T> next() {
+      try {
+        return CompletableFuture.completedFuture(input.nextValue());
+      } catch (Throwable e) {
+        return CompletableFutures.failedFuture(e);
+      }
+    }
+  }
+
+  private static final class ComposingStreamingResponse<I, O> extends StreamingResponse<O> {
+    private final StreamingResponse<I> input;
+    private final Function<? super I, ? extends CompletableFuture<O>> transform;
+
+    private ComposingStreamingResponse(
+        StreamingResponse<I> input, Function<? super I, ? extends CompletableFuture<O>> transform) {
+      this.input = requireNonNull(input);
+      this.transform = requireNonNull(transform);
+    }
+
+    @Override
+    public boolean hasNext() {
+      return input.hasNext();
+    }
+
+    @Override
+    public CompletableFuture<O> next() {
+      return input.next().thenCompose(transform);
+    }
+  }
+
+  private static final class AsyncResponseQueue {
+    private static final String CHUNK_SEPARATOR = "\r\n";
+
+    private final ChunkedOutput<ResultOrError> sink;
+
+    // tail is the end of a linked list of completable futures. The futures are tied together by
+    // allOf completion handles. For example, let's say we push 3 futures into the queue:
+    //
+    // 1. tail = (null) // empty queue
+    // 2. tail = then(allOf((null), f_1), write(f_1)) // -> f_1
+    // 3. tail = then(allOf(then(allOf((null), f_1), write(f_1)), f_2), write(f_2)) // -> f_1 -> f_2
+    //
+    // The chaining of the futures guarantees that the write(f) callbacks are called in the order
+    // that the futures got pushed into the queue. Also, due to the way allOf is implemented, it
+    // makes sure that, as soon as f_0..i is completed, the whole
+    // then(allOf((f_i-1), f_i), write(f_i)) monad is made available for garbage collection.
```

Comment:
fancy. i haven't completely checked what's in this file for correctness, but it's at least approximately set up nicely and you're thinking about what needs to be thought about.

Reply:
This is the bit that guarantees the ordering. i.e. `write` callback will be called in the order the futures are appended in the queue.